service/storage_service.cc

```diff
@@ -683,7 +685,18 @@ void storage_service::join_token_ring(int delay) {
                 throw std::runtime_error("This node is already a member of the token ring; bootstrap aborted. (If replacing a dead node, remove the old one from the ring first.)");
             }
             set_mode(mode::JOINING, "getting bootstrap token", true);
-            _bootstrap_tokens = boot_strapper::get_bootstrap_tokens(_token_metadata, _db.local());
+            if (resume_bootstrap) {
+                _bootstrap_tokens = db::system_keyspace::get_saved_tokens().get0();
```

Comment:
Is there a chance that you could split this patch into those modifying specific ops (e.g. bootstrapping)?

Why we didn't use `get_saved_tokens()` before? This seems to be unrelated to switching to repair-based ops.

Reply:
Do you mean splitting "storage_service: Switch to repair based node operation" into 5 patches each enables one operation?

Using get_saved_tokens() is related to switching repair-based bootstrap operation, because without it we could not actually resume from the previous failed bootstrap due to different tokens used. 

We are not using it before, because it does not matter if we use the same tokens or not since the bootstrapping node will going to stream all the ranges again anyway.

Reply:
> Do you mean splitting "storage_service: Switch to repair based node operation" into 5 patches each enables one operation?

Yes. This in particular would allow you have a shorter and more specific patch description for each modified operation and this would make it much easier to read (and review).

However, from the patch perspective my main complain is about the lines in the "bootstrapping" part that are beyond a mechanical change. 

Please, see my comments on that below.

> 
> Using get_saved_tokens() is related to switching repair-based bootstrap operation, because without it we could not actually resume from the previous failed bootstrap due to different tokens used.
> 
> We are not using it before, because it does not matter if we use the same tokens or not since the bootstrapping node will going to stream all the ranges again anyway.

I see. However it could make sense to add just this ("using `get_saved_tokens()`") as a separate patch because this has a potential for breaking things. Then you can add the rest of changes as a single patch if you want (while I think it would still be nicer to break them into separate operations).

The point is that this way the series would be better "git bisectable" and it would make possible to differentiate breakages coming from using `get_saved_tokens()` and those coming from switching to repair based ops.

However I do see your point that using `get_saved_tokens()` may be seen as an integral part of "switching to repair based ops" patch.

So, your call.

Reply:
Thanks. I will think about it. There are more requests from Avi to enable only some of the operations.